dcs/src/main/java/org/trafodion/dcs/server/ServerManager.java

```diff
@@ -414,32 +431,40 @@ public Boolean call() throws Exception {
                                                              // finish
                 if (f != null) {
                     Integer result = f.get();
+                    LOG.debug("Server handler [" + instance + ":" + result + "] finished");
                     int childInstance = result.intValue();
                     // get the node id
-                    boolean isRunning = serverHandlers[childInstance-1].serverMonitor.isPidRunning();
-                    String nid = serverHandlers[childInstance-1].serverMonitor.nid;
-                    String pid = serverHandlers[childInstance-1].serverMonitor.pid; 
-                    serverHandlers[childInstance-1] = null;
-                    LOG.debug("Server handler [" + instance + ":" + result
-                            + "] finished, restarting");
-                    if (isRunning)
-                        LOG.info("mxosrvr " + nid + "," + pid + " still running");
-                    else
-                        LOG.info("mxosrvr " + nid + "," + pid + " exited, restarting");
+                    boolean isRunning = serverHandlers[childInstance - 1].serverMonitor.monitor();
+                    String nid = serverHandlers[childInstance - 1].serverMonitor.nid;
+                    String pid = serverHandlers[childInstance - 1].serverMonitor.pid;
+                    int restartAttempts = serverHandlers[childInstance - 1].getRestartAttempts();
+
+                    serverHandlers[childInstance - 1] = null;
                     retryCounter = retryCounterFactory.create();
                     while (!isTrafodionRunning(nid)) {
-                       if (!retryCounter.shouldRetry()) {
-                          throw new IOException("Node " + nid + " is not Up");
-                       } else {
-                           retryCounter.sleepUntilNextRetry();
-                           retryCounter.useRetry();
-                       }
-                   }
-                   serverHandlers[childInstance-1] = new ServerHandler(childInstance);
-                   completionService.submit(serverHandlers[childInstance-1]);
+                        if (!retryCounter.shouldRetry()) {
```

Comment:
Agree with Selva, please keep retryCounterFactory  and get rid of new retry logic.